src/protobuf-net.Core/Serializers/MapSerializer.cs

```diff
@@ -201,4 +207,54 @@ internal override void Write(ref ProtoWriter.State state, int fieldNumber, WireT
             }
         }
     }
+
+    class IReadOnlyDictionarySerializer<TKey, TValue> : MapSerializer<IReadOnlyDictionary<TKey, TValue>, TKey, TValue>
+    {
+        protected override IReadOnlyDictionary<TKey, TValue> Initialize(IReadOnlyDictionary<TKey, TValue> values, ISerializationContext context)
+            => values ?? new Dictionary<TKey, TValue>();
+
+        protected override IReadOnlyDictionary<TKey, TValue> Clear(IReadOnlyDictionary<TKey, TValue> values, ISerializationContext context)
+        {
+            if (values is Dictionary<TKey, TValue> target)
+            {
+                target.Clear();
+                return target;
+            }
+            return new Dictionary<TKey, TValue>();
+        }
+
+        protected override IReadOnlyDictionary<TKey, TValue> AddRange(IReadOnlyDictionary<TKey, TValue> values, ref ArraySegment<KeyValuePair<TKey, TValue>> newValues, ISerializationContext context)
+        {
+            if (values is not Dictionary<TKey, TValue> target)
```

Comment:
Thought: maybe we just need to check `is not IDictionary<TKey, TValue> dict || dict.IsReadOnly` ?

Reply:
ack